asmcomp/polling.ml

```diff
@@ -16,182 +17,226 @@ open Mach
-  | Iswitch (_, acts) ->
-      (Array.exists check_path acts) || (check_path i.next)
-  | Icatch (_, handlers, body) ->
-     (check_path body)
-     ||
-      (List.exists check_path (List.map (fun (_, handler_code) ->
-        handler_code) handlers)) || (check_path i.next)
-  | Itrywith (body, handler) ->
-      (check_path body) || (check_path handler) || (check_path i.next)
-  | Iop (Itailcall_ind) -> true
-  | Iop (Itailcall_imm { func; _ }) ->
-    if (String.Set.mem func future_funcnames) then
-      (* this means we have a call to a function that might be a self call
-         or a call to a future function (which won't have a poll) *)
-      true
+(* Detection of recursive handlers that are not guaranteed to poll
+   at every loop iteration. *)
+
+(* The result of the analysis is a mapping from handlers H
+   (= loop heads) to Booleans b and polling decisions.
+
+   b is true if every path starting from H goes through an Ialloc,
+   Ipoll, Ireturn, Itailcall_ind or Itailcall_imm instruction.
+   In this case, we say that H is "safe".
+   b is also true if we decide to insert a poll at the head of H.
+
+   (Iraise is treated like Ireturn if not handled locally, or like
+    a branch to the nearest enclosing exception handler.)
+
+   b is false, therefore, if starting from H we can loop infinitely
+   without crossing an Ialloc or Ipoll instruction.
+   In this case, we say that H is "unsafe".
+*)
+
+(* The analysis is a backward dataflow analysis starting from false,
+   using && (Boolean "and") as the join operator,
+   and with the following transfer function:
+
+   TRANSF(Ialloc | Ipoll | Itailcall_ind | Itailcall_imm _ | Ireturn) = true
+   TRANSF(all other operations, x) = x
+*)
+
+(* We modify the code during the analysis in the following way:
+   As soon as we find a recursive catch with some unsafe handlers
+   we mark one of them for poll insertion, then we resume the analysis
+   to determine whether there are still some unsafe handlers.
+
+   Note that this means the analysis must return true in the end because
+   it will have inserted polls in all unsafe loops.
+*)
+
+type handler_status = AddPoll | Polls | MayLoop
+
+let polled_loops_analysis funbody =
+  let handlers : (int, handler_status) Hashtbl.t = Hashtbl.create 20 in
+  let get_handler n =
+    match Hashtbl.find_opt handlers n with
+    | Some (AddPoll | Polls) -> true
+    | Some (MayLoop) | None -> false
+  in
+  let set_handler n b =
+    match Hashtbl.find_opt handlers n with
+    | Some AddPoll -> ()
+    | _ -> Hashtbl.replace handlers n (if b then Polls else MayLoop)
+  in
+  let set_add_poll n = Hashtbl.replace handlers n AddPoll in
+  let get_add_poll n =
+    match Hashtbl.find_opt handlers n with
+    | Some AddPoll -> true
+    | _ -> false
+  in
+  let rec before i end_ exn =
+    match i.desc with
+    | Iend -> end_
+    | Iop (Ialloc _ | Ipoll _ | Itailcall_ind | Itailcall_imm _) -> true
```

Comment:
I'm afraid this "shortcut" is not right.  By not computing `before i.next end_ exn`, we omit to analyze loops that occur after the `Ialloc` instruction.  These loops will have the default "false" abstract value, hence a polling point will always be inserted in these loops.  

Using a pedestrian dataflow engine that always goes over all the instructions (#10404) may be a good idea, after all.